code/modules/research/mechanic/component_exchanger.dm

```diff
@@ -0,0 +1,254 @@
+
+	if(interactoption == "Replace Component")
+
+		user.visible_message("<span class='notice'>[user] carefully fits \the [src] into \the [M] as it rattles and starts remplacing components.</span>", \
+		"<span class='notice'>\The [src]'s HUD flashes, a message appears stating it has started scanning and replacing \the [M]'s components.</span>")
+
+		for(var/obj/item/weapon/stock_parts/P in M.component_parts)
+			if(!Adjacent(M, user)) //Make sure the user doesn't move
+				user << "<span class='warning'>A blue screen suddenly flashes on \the [src]'s HUD. It appears the critical failure was caused by suddenly yanking it out of \the [M]'s maintenance hatch.</span>"
+				return
+			//Yes, an istype list. We don't have helpers for this, and this coder is not that sharp
+			if(istype(P, /obj/item/weapon/stock_parts/capacitor))
+				for(var/obj/item/weapon/stock_parts/capacitor/R in componentstorage)
+					if(R.rating > P.rating && P.loc != src) //Kind of a hack, but makes sure we don't replace components that already were
+						sleep(10) //One second per component
+						perform_indiv_replace(P, R, M)
+						//Do not break in case we find even better
+			if(istype(P, /obj/item/weapon/stock_parts/scanning_module))
+				for(var/obj/item/weapon/stock_parts/scanning_module/R in componentstorage)
+					if(R.rating > P.rating && P.loc != src)
+						sleep(10) //One second per component
+						perform_indiv_replace(P, R, M)
+			if(istype(P, /obj/item/weapon/stock_parts/manipulator))
+				for(var/obj/item/weapon/stock_parts/manipulator/R in componentstorage)
+					if(R.rating > P.rating && P.loc != src)
+						sleep(10) //One second per component
+						perform_indiv_replace(P, R, M)
+			if(istype(P, /obj/item/weapon/stock_parts/micro_laser))
+				for(var/obj/item/weapon/stock_parts/micro_laser/R in componentstorage)
+					if(R.rating > P.rating && P.loc != src)
+						sleep(10) //One second per component
+						perform_indiv_replace(P, R, M)
+			if(istype(P, /obj/item/weapon/stock_parts/matter_bin))
+				for(var/obj/item/weapon/stock_parts/matter_bin/R in componentstorage)
+					if(R.rating > P.rating && P.loc != src)
+						sleep(10) //One second per component
+						perform_indiv_replace(P, R, M)
+			//Good thing there's only a few stock parts types
+
+		M.RefreshParts()
+		user.visible_message("<span class='notice'>[user] pulls \the [src] out of \the [M] as it finishes remplacing components.</span>", \
+		"<span class='notice'>You pull \the [src] out of \the [M] as a message flashes on its HUD stating it has finished remplacing components and will return to the input screen shortly.</span>")
+
+		spawn(30)
+			component_interaction(user, M)
+
+//So we don't copy the same thing a thousand fucking times
+/obj/item/device/component_exchanger/proc/perform_indiv_replace(var/obj/item/weapon/stock_parts/P, var/obj/item/weapon/stock_parts/R, var/obj/machinery/M)
```

Comment:
I tried it with oldpart and newpart, and it was a goddamn nightmare so I just reverted to old behavior

It's readable as long as you follow the comments. The part where we deal with the old part (P) and the new part (R) are clearly outlined